rest/authorization-server/src/main/java/org/eclipse/sw360/rest/authserver/security/customheaderauth/Sw360CustomHeaderAuthenticationProvider.java

```diff
@@ -0,0 +1,178 @@
+ * let us know about the authentication.
+ *
+ * In addition it is special because it calculates the granted authorities for
+ * the user depending on the user's authorities given by his groups and the
+ * client's scopes. The result will be the intersection between these two lists.
+ * Of course this is only done for an oauth request and not for normal ones
+ * (that have nothing to do with clients). And in fact he uses for this task the
+ * {@link Sw360UserAndClientAuthoritiesMerger}.
+ */
+public class Sw360CustomHeaderAuthenticationProvider implements AuthenticationProvider {
+
+    private final Logger log = Logger.getLogger(this.getClass());
+
+    @Value("${security.customheader.headername.intermediateauthstore:#{null}}")
+    private String customHeaderHeadernameIntermediateAuthStore;
+
+    @Autowired
+    private Sw360UserDetailsProvider sw360CustomHeaderUserDetailsProvider;
+
+    @Autowired
+    private ClientDetailsService clientDetailsService;
+
+    @Autowired
+    private Sw360UserAndClientAuthoritiesMerger sw360UserAndClientAuthoritiesMerger;
+
+    private boolean active;
+
+    @PostConstruct
+    public void postSw360CustomHeaderAuthenticationProviderConstruction() {
+        if (StringUtils.isEmpty(customHeaderHeadernameIntermediateAuthStore)) {
+            log.warn("AuthenticationProvider is NOT active! Some configuration is missing. Needed config keys:\n"
+                    + "- security.customheader.headername.intermediateauthstore");
+            active = false;
+        } else {
+            log.info("AuthenticationProvider is active!");
+            active = true;
+        }
+    }
+
+    @Override
+    public boolean supports(Class<?> authentication) {
+        return active
+                && (authentication.equals(UsernamePasswordAuthenticationToken.class) || authentication.equals(PreAuthenticatedAuthenticationToken.class));
+    }
+
+    @Override
+    public Authentication authenticate(Authentication authentication) throws AuthenticationException {
+        // check if the marker header of our filter is available
+        if (authentication.getDetails() instanceof Map<?, ?>
+                && ((Map<?, ?>) authentication.getDetails()).containsKey(customHeaderHeadernameIntermediateAuthStore)) {
+            Map<?, ?> authDetails = ((Map<?, ?>) authentication.getDetails());
+
+            // get user details
+            String email = (String) authentication.getPrincipal();
+            Object externalIds = authDetails.get(customHeaderHeadernameIntermediateAuthStore);
+            String externalId;
+            if (externalIds != null && externalIds instanceof String[]) {
+                externalId = ((String[]) externalIds)[0];
+            } else {
+                externalId = (String) externalIds;
+            }
+            User userDetails = sw360CustomHeaderUserDetailsProvider.provideUserDetails(email, externalId);
+
+            List<GrantedAuthority> grantedAuthorities = new ArrayList<>();
```

Comment:
Don't get the second question?

Reply:
> Don't get the second question?

At this point I was struggling to understand the differences between my proposed function `calculateGrantedAuthorities` and `sw360UserAndClientAuthoritiesMerger.mergeAuthoritiesOf` since they do in one case the same or might have the same function signature.

I think that will be fixed once the code duplication is resolved.

Reply:
Ack, let's after my changes.